docs: [no-floating-promises] clarify that void does not resolve promises (#9949)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/no-floating-promises.mdx

Floating Promises can cause several issues, such as improperly sequenced operations, ignored Promise rejections, and more.

This rule reports when a Promise is created and not properly handled.

> Valid ways of silencing a Promise-valued statement include:

Reviewer: What do you think of something more along these lines?

```suggestion
This rule will report Promise-valued statements that are not treated in one of the following ways:
```

The idea I'm getting at is to just say what's reported or not reported, and dodge unpacking every scenario that does or doesn't lead to unhandled rejection, and why the rule cannot report in 1:1 correspondence with those scenarios.

Author: Yeah, I chose to just use "silencing" because of the issue I've opened before this PR.
But yeah, I like this.

- `await`ing it (❌ unhandled)
- `return`ing it (❌ unhandled)

> - [`void`ing it](#ignorevoid) (❌ unhandled)

Reviewer: (praise) Good call on adding the `#ignorevoid` link!

> - [`void`ing it](#ignorevoid) (❌ unhandled)

Reviewer: Maybe it would be good to call out here 
```suggestion
- [explicitly marking the promise as ignored by `void`ing it](#ignorevoid)
```

What do you think?

(and i personally would move it to the end of the list since it's an outlier in that it is just a way to disable the rule, not actually handle the promise, but up to you)

Author: I'm neutral to this suggestion. I mean, using "`void`ing" is coherent with "`await`ing" and "`return`ing" wording, so I'm not actually sure.

About moving to the bottom, this is perfectly fine. I like it.

- Calling its `.then()` with two arguments (✅ handled)

> - Calling its `.catch()` with one argument (✅ handled)

Reviewer: I understand why you'd want to have these `(❌ unhandled)`/`(✅ handled)` indicators, but I'm not sure they're an improvement here, without a deeper dive into unhandled rejections. 

The reason I say this is that code like `promise.catch(() => { throw new Error('unhandled!') });` _is_ still unhandled, but (intentionally) not flagged by the rule.

Furthermore, `await`ing or `return`ing promises _do_ prevent a unhandled rejections (assuming the caller is handled appropriately).

For now, my thought would be to leave these markers off (but maybe this can be revisited when we do futher docs improvements for the promise rules, which is already on our redar).

Author: I see your point. I'd ask you why this is intentional, but I guess that would be pointless.

`await`ing and `return`ing do not prevent unhandled rejections but delegate the handling to, respectively a `try-catch` and the parent function.

As you surely have seen, these markers are meant also for the `ignoreVoid` section warning. So I wonder if there's another way to include such markers in a softer way, maybe by preparing it for the change you mentioned...

Author: I've removed the markers as I've improved the message in `ignoreVoid`.
Let me know if it is okay to you

Reviewer: > I'd ask you why this is intentional, but I guess that would be pointless

Hardly! 🙂 The short version is just that it's impossible to know that a function won't throw in JS, so, in reality, `.then()` and `.catch()` can never guaranteeably prevent a floating promise. 

For example, `promise.catch(() => maybeThrows())` can never be known not to throw, so _technically_ we could mark it as unhandled. But then, _everything_ would be unhandled! So instead, we say, `promise.catch(() => maybeThrows())` _looks_ like it's handled, so the rule should not report.

Similarly, for `promise.then(() => onFulfilled(), () => onRejected())`, we know that `onFulfilled` _could_ throw, and isn't any different than `promise.then(() => onFulFilled())`, which we _would_ mark as unhandled. However, because the `onRejected` handler is provided, we say that the usage looks like the developer was aware of rejections when they wrote the code, and, therefore, it it _looks_ handled. 

So, no-floating-promises really becomes about putting up reasonable guardrails, rather than attempting to make guarantees.

There are quite a few issues that discuss these tradeoffs.

Author: The fact it "looks like it handled" could be a thing worth to be specified in a new paragraph about handling promises.

- `Promise.any()`
- `Promise.race()`

Refer to [Handling Promises Rejections](#handling-promises-rejections) below for more information about this rule.

> 

Reviewer 2: I don't see any real value to including this. The page is tall; people will know that scrolling down will give them more information.

Author: I wouldn't make this assumption. I'm speaking for myself, but I often try to scan the page fastly and perform a search to get what I want. I don't read everything.

However, this is fine if I move the paragraph below the list, as suggested in the above.

Author: I moved the paragraph up and removed the phrase. Let me know.

- `Promise.race()`

:::tip

> `no-floating-promises` only detects unhandled Promise _statements_ and voided ones.

Author: Thanks for the explanation. I'm going to revert it

- `Promise.race()`

:::tip
`no-floating-promises` only detects apparently unhandled Promise _statements_.
See [`no-misused-promises`](./no-misused-promises.mdx) for detecting code that provides Promises to _logical_ locations such as if statements.

> See [_Using promises (error handling) on MDN_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#error_handling) for a detailed writeup on promise error-handling.

Reviewer 2: ```suggestion
See [_Using promises (error handling) on MDN_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#error_handling) for a detailed writeup on Promise error-handling.
```

Author: I think this is not mine, but I still accepted it

</TabItem>
</Tabs>

> ## Handling Promises rejections

Reviewer: Oh, sorry. I feel I may have led you in the wrong direction here. I'm thinking, for now, for this PR, let's not add this section (sorry for the time you spent on this!). 

But it _would_ be in scope to prominently link to the MDN error-handling writeup, maybe by a simple `See [the MDN docs](link) for a detailed writeup on promise error-handling.` right below the list of ways to satisfy the rule?

Author: I love writing things and explanations for people to read.

I studied this paragraph with the idea of it linking to MDN, not replacing it, but also to share the knowledge we discussed in here and adding some values related to this rule.

I think that removing such paragraph and reducing it to be just a link, would almost completely void (_pun intended_) the purpose of this PR, which I guess could be closed then.

We would be going to remove the whole concept of "looks like to be handled" and "guardrails" you highlighted before in this discussion, which are two concepts that I think are worth to me mentioned in terms of mental model.

I can move it below the list, for sure, and I can tune the paragraph a bit.

However, I'm of the idea I already spent too much energy on this hoping it would be okay and reducing it to a link makes me feel I wasted my time and energy. I hope you understand how I'm feeling after this comment.

Author: I tried to rework the paragraph a little bit and moved it up in the discussion. Let me know how it feels now.

Reviewer: @alexandercerutti, I hear you. But, it's deliberate that the issue was scoped narrowly to be essentially about the `void` wording only, see https://github.com/typescript-eslint/typescript-eslint/issues/9947#issuecomment-2334294112. We have #8088 and #8517 already to address the larger scale efforts of improving the promise docs, deep-diving on the technical details, and ensuring that the various promise rules' docs are coherent with each other, rather than just this page alone.

So, we all agree that the deep-dive info on promise rejections needs to exist, but, unfortunately, I need to insist that this particular PR is not the time and place to make it happen, sorry.

We'll certainly keep this discussion in mind, though for when it does happen in the linked issues!

Best, Kirk

This option, which is `true` by default, allows you to stop the rule reporting promises consumed with the [`void` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void).
This can be a good way to explicitly mark a promise as intentionally not awaited.

:::warning
Voiding a Promise doesn't handle it or change the runtime behavior. The outcome is just ignored, like disabling the rule with a

```
// eslint-disable-next-line @typescript-eslint/no-floating-promises
```

> Such Promise rejection will still be unhandled, as explained in the above section [_Handling Promises Rejections_](#handling-promises-rejections).

Reviewer: ```suggestion
Such Promise rejections will still be unhandled.
```

This option, which is `true` by default, allows you to stop the rule reporting promises consumed with the [`void` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void).
This can be a good way to explicitly mark a promise as intentionally not awaited.

:::warning
Voiding a Promise doesn't handle it or change the runtime behavior. The outcome is just ignored, like disabling the rule with a

```
// eslint-disable-next-line @typescript-eslint/no-floating-promises
```

> Such Promise rejection will still lead to an [`unhandledrejection` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event)

Reviewer: Continuing from https://github.com/typescript-eslint/typescript-eslint/pull/9949#discussion_r1748907511,

> ... my idea was to exactly represent what happend to me and what I described in the issue (the code example).

Right, exactly. But the details of exactly what happened to you are a little specific imo since it's dependent on node vs web, and on whether a global promise rejection handler was or was not set up.

> So, I think it is important to include a link to mdn page about `unhandledrejection`, which can explain surely better what is it.
> 
> About the comment above, I think could be interesting to give a hint about how can a person handle a promise.

To be clear, I definitely agree with putting this info on this page! 🙂  I'm just thinking it isn't _specific_ to the `void` operator - it's essential knowledge whether or not you're using the `ignoreVoid` option - so I'd prefer it to be in the general info on handling promises, rather than in this particular warning block (which I wouldn't bother reading if i had the `ignoreVoid` option disabled).

Note that the MDN page link (albeit different section) is included already, down in the "Further Reading" section. Maybe we can find a better place to link the "error handling" section in the intro?

What do you think?

Author: > Right, exactly. But the details of exactly what happened to you are a little specific imo since it's dependent on node vs web, and on whether a global promise rejection handler was or was not set up.

I just tested the difference between node and web. I see some differences I'm going to document here.

 - `unhandledrejection` (web) vs `unhandledRejection` (node)
 - `window` vs `process`
 - if the listener is set up, error log still prints on web, but doesn't on node

So, I see your point.

> so I'd prefer it to be in the general info on handling promises, rather than in this particular warning block (which I wouldn't bother reading if i had the ignoreVoid option disabled).

I agree. The link to `ignoreVoid` was serving that exact purpose, but it might not be enough. There is no section about handling promises, right?

The fact the MDN link is on the end section might not help making it being read. Personally I don't ever read to the end of a document. This behavior might be wrong, but I organize my document sections to give the important resources first.

I'm going to try to put down a few lines

This option, which is `true` by default, allows you to stop the rule reporting promises consumed with void operator.
This can be a good way to explicitly mark a promise as intentionally not awaited.

> :::warning
> Voiding a Promise doesn't handle it, but just ignores the outcome. Such Promise rejection will still lead to an
> [`unhandledrejection` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event)
> being fired and error logs being printed. Refer to the list at the top of the page for the possible ways to handle a promise.
> :::

Reviewer: (praise) I quite like that you've used the `:::warning` block here 🙂 

... but I feel like the wording wanders a bit off topic. What do you think of something like this?

```suggestion
:::warning
Voiding a promise doesn't handle it or change its behavior in any way. It is equivalent to an `// eslint-disable-next-line @typescript-eslint/no-floating-promises` comment.
:::
```

Author: Quoting the `disable-next-line` fact is great and I can add it. However, my idea was to exactly represent what happend to me and what I described in the issue (the code example).

So, I think it is important to include a link to mdn page about `unhandledrejection`, which can explain surely better what is it.

About the comment above, I think could be interesting to give a hint about how can a person handle a promise.

This option, which is `true` by default, allows you to stop the rule reporting promises consumed with the [`void` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void).
This can be a good way to explicitly mark a promise as intentionally not awaited.

:::warning

> Voiding a Promise doesn't handle it or change the runtime behavior. The outcome is just ignored, like disabling the rule with a
> 
> ```
> // eslint-disable-next-line @typescript-eslint/no-floating-promises
> ```
> 
> Such Promise rejections will still be unhandled.

Reviewer 2: Lots of vertical space for a point that could be made with a link, no?

```suggestion
Voiding a Promise doesn't handle it or change the runtime behavior.
The outcome is just ignored, like disabling the rule with an [ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1).
Such Promise rejections will still be unhandled.
```

Author: Yes, why not
